### ISSUE TITLE
Multiple @ManyToOne relationships using the same join column doesn't work

### ISSUE THREAD

**Issue by PachowStudios** [typeorm/typeorm#6582]
**Issue type:**

[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[x] `0.2.24` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

Given the following:
```ts
@Entity()
export class Foo {
  @Field(() => ID)
  @PrimaryGeneratedColumn('uuid')
  id!: string;

  @Field(() => ID)
  @Column()
  thingId!: string;

  @ManyToOne(() => ThingA)
  @JoinColumn({ name: 'thing_id' })
  thingA?: ThingA;

  @ManyToOne(() => ThingB)
  @JoinColumn({ name: 'thing_id' })
  thingB?: ThingB;
}
```

Doing a join like this:
```ts
Foo.createQueryBuilder('foo')
  .innerJoin('foo.thingA', 'thingA')
  .getMany();
```

Fails with the following error:
`QueryFailedError: column thingA.thingId does not exist`

This is the problematic part of the resulting SQL:
```sql
INNER JOIN "thingA" "thingA" ON thing_a.thingId = "foo"."thing_id"
```

The proper SQL would look like this:
```sql
INNER JOIN "thingA" "thingA" ON "thing_a"."id" = "foo"."thing_id"
```

**Here's where things get weird**
It works as soon as I remove the `thingB` column. Even though I'm not joining on `thingB` in my query.

Adding `referencedColumnName: 'id'` to `thingA`'s `@JoinColumn` doesn't help either.

I also discovered that if I swap the order that the relations `thingA` and `thingB` are declared, it works (either that or if I reference `thingB` instead of `thingA`).

I think the problem code is `RelationJoinColumnBuilder.collectColumns` [here](https://github.com/typeorm/typeorm/blob/master/src/metadata-builder/RelationJoinColumnBuilder.ts#L168). The issue appears to be that when registering a join column, it overwrites whatever one was previously assigned.

**Comment by jameswpierce** [comment]
This is a limitation of relational databases. You are only creating one join column, which can only store the id of one item. In order to store the ids of the separate 'things' you will need to name the join columns differently, i.e. 'thing_a_id', and 'thing_b_id'. Then each row in your table will be able to store references to `thing_a` and `thing_b` separately.

**Comment by PachowStudios** [comment]
They aren't separate things though. They all share the same id.

**Comment by jameswpierce** [comment]
Ok! I am not sure what the use case for that would be. If they are the same thing then you shouldn't need separate relationships! 

If you can elaborate more on what you are trying to do by having two class properties that point to the same related object, maybe someone can point you in the right direction, but it sounds like something that may need to be handled in a different context than the definition of your model's schema.

**Comment by PachowStudios** [comment]
I'm representing a polymorphic relationship. For example base entity Identity could have child tables Person and/or Student.

**Comment by imnotjames** [comment]
Duplicate of #1668

**Comment by imnotjames** [comment]
You've done a ton of really awesome debugging.  Had I found this earlier today you'd have saved me a bunch of time. 😆 

This is a long term known issue and I'm currently noodling over in the related issue about how to resolve it.

I'll be closing this out but definitely interested in hearing any more you might have in the matter.  This might require a bit of a rework on the way relations are currently modeled a bit.

**Comment by bashleigh** [comment]
I've built a basic implementation for polymorphic children and parents, I did write this a long time ago though and it really, really, really needs a clean up. I've implemented this soluion for 2 production projects and it's worked. However I've now deployed this to npm as `typeorm-polymorphic` if anyone is interested or needs this functionality. Currently adding tests so I can clean up and improve without effect the existing functionality! I do also need to add query builder methods which would be really, really cool. However I do feels it's a little bit of a hacky solution but it does the job!
